feat(material-experimental/mdc-list): add test harness

mmalerba opened the pull request:
Note: This PR is based on #20510, rebased with fixes for a few things that changed since it was created.

jelbourn commented on src/material-experimental/mdc-list/list.scss:
```diff
@@ -50,16 +50,22 @@
+  // If the avatar is displayed in the graphic section of a list-item (i.e. at the start),
+  // we override the default "graphic" horizontal spacing that is designed for icons, to
+  // work for actual avatar images. The avatar by default is used within the `graphic` section,
+  // but could be also outside in case of a selection-list option with a checkbox at start.
```
I don't totally follow this comment- what's the practical difference between what we support and what MDC supports here?

mmalerba replied:
MDC assumes that the list as a whole will either show avatars (larger images) or icons (smaller images). However, we allow avatar vs icon to be decided on a per-item basis. Therefore we can't rely on their styles which are something like `.mdc-list.mdc-list--avatar-list .mdc-list-item ... {}`. Instead we add similar styles of are own.

This part specifically is dealing with avatars that are marked as `.mdc-list-item__graphic`. MDC list items have 2 content slots for icons:
- `.mdc-list-item__graphic` is for an icon that comes before the text
- `.mdc-list-item__meta` if for an icon that comes after the text

Their styles targeted at `.mdc-list--avatar-list` would normally override these margins to look right for avatars rather than icons, but since we can't use those styles we're doing the same thing manually here.

Most of this is covered by various existing comments in the file, but I'll see if I can organize it to make it a bit more clear.

jelbourn commented on src/material-experimental/mdc-list/list-styling.ts:
```diff
@@ -6,15 +6,39 @@
+
+/**
+ * Directive whose purpose is to add the MDC list-item `graphic` or `meta`
+ * class depending on checkbox position in list options.
```
Could this comment be expanded to elaborate more on the context here? (i.e. why this is necessary)

jelbourn commented on src/material-experimental/mdc-list/testing/list-item-harness-base.ts:
```diff
@@ -0,0 +1,89 @@
+
+/** Selectors for the various list item sections that may contain user content. */
+export const enum MatListItemSection {
+  CONTENT = '.mdc-list-item__text',
```
Add a comment here about types of section that would be added later?

jelbourn reviewed (approved):
LGTM

mmalerba commented:
@jelbourn addressed your feedback in a new commit. This PR is marked as `preserve commits` so I'll manually merge it into the previous commit once you OK it.
